File: clock.rb
# create font of characters:
num_0 = [" _ _  ",
    "|   | ",
    "|_ _| "]
num_1 = ["   ",
    "/| ",
    " | "]
num_2 = [" _ _  ",
    " _ _| ",
    "|_ _  "]
num_3 = ["_ _  ",
    "_ _| ",
    "_ _| "]
num_4 = ["      ",
    "|_ _| ",
    "    | "]
num_5 = [" _ _  ",
    "|_ _  ",
    " _ _| "]
num_6 = [" _ _  ",
    "|_ _  ",
    "|_ _| "]
num_7 = ["_ _ ",
    "  / ",
    " /  "]
num_8 = [" _ _  ",
    "|_ _| ",
    "|_ _| "]
num_9 = [" _ _  ",
    "|_ _| ",
    " _ _| "]       
symbol = ["    ",
     " *  ",
     " *  "]

font = [num_0, num_1, num_2, num_3, num_4, num_5, num_6, num_7, num_8, num_9, symbol]
# in: datetime of OS
# out: string "15:30:22"
def read_system_clock()
    full_time_str =Time.now.to_s
    puts full_time_str
    hhmmss=""
    for i in 11..18 do
        hhmmss+=full_time_str[i]
    end
    return hhmmss
end 


#in : string "15:30:22" and font
#out : [[],[],[],...]
def convert_characters_to_fonts(time_str,font)
    out=[]
    for i in 0..(time_str.length-1) do
        character=time_str[i]
        converted_font=convert_1_character_to_1_font(character,font)
        out.push(converted_font)
    end
    return out
end

# int:
# out: ["row_1", "row_2", ... ]
def merge_arr (arr)
    merge=["","","","","","","","","",""]
    for i in 0..(arr[0].length-1) do    
        for j in 0..(arr.length-1) do
            font=arr[j]
            merge[i]+=font[i]
        end
    end
    return merge
end

# int: character & font
# out: [] tuong ung voi 1 font
def convert_1_character_to_1_font(character,font)
    if character=="0"
        return font[0]
    end
    if character=="1"
        return font[1]
    end
    if character=="2"
        return font[2]
    end
    if character=="3"
        return font[3]
    end
    if character=="4"
        return font[4]
    end
    if character=="5"
        return font[5]
    end
    if character=="6"
        return font[6]
    end
    if character=="7"
        return font[7]
    end
    if character=="8"
        return font[8]
    end
    if character=="9"
        return font[9]
    end
    if character==":"
        return font[10]
    end
end
while true do
    system("clear")
    time_str=read_system_clock()
    converted=convert_characters_to_fonts(time_str,font)
    merged=merge_arr(converted)
    puts merged
    sleep(1)
end

File: clock.go
package main
import "fmt"
import "time"
import "strings"
import "os"
import "os/exec"
//create font of characters :
var font=[][]string{
	[]string{" _ _  ",
	         "|   | ",
			 "|_ _| "},
	[]string{"   ",
	         "/| ",
			 " | "},
	[]string{" _ _  ",
			 " _ _| ",
			 "|_ _  "},
	[]string{"_ _  ",
	         "_ _| ",
			 "_ _| "},
	[]string{"      ",
	         "|_ _| ",
			 "    | "},
	[]string{" _ _  ",
	         "|_ _  ",
			 " _ _| "},
	[]string{" _ _  ",
	         "|_ _  ",
			 "|_ _| "},
	[]string{"_ _ ",
	         "  / ",
			 " /  "},
	[]string{" _ _  ",
	         "|_ _| ",
			 "|_ _| "},
	[]string{" _ _  ",
	         "|_ _| ",
			 " _ _| "},
	[]string{"    ",
	         " *  ",
			 " *  "},	
}

// in: datetime of OS
// out: string "07:35:43"
func read_system_clock()string{
	full_time := time.Now()
	fmt.Println(full_time)
	hms:=full_time.String()
	hhmmss:=string(hms[11:19])
	fmt.Println(hhmmss)
	return hhmmss
}
// int: character & font
// out: [] tuong ung voi 1 font
func convert_1_character_to_1_font(character string,font [][]string)[]string{
	var str="0123456789:"
	s:=strings.Split(str,"")
	var f []string
	if character==s[0]{
		f=font[0]
	}
	if character==s[1]{
		f=font[1]
	}
	if character==s[2]{
		f=font[2]
	}
	if character==s[3]{
		f=font[3]
	}
	if character==s[4]{
		f=font[4]
	}
	if character==s[5]{
		f=font[5]
	}
	if character==s[6]{
		f=font[6]
	}
	if character==s[7]{
		f=font[7]
	}
	if character==s[8]{
		f=font[8]
	}
	if character==s[9]{
		f=font[9]
	}
	if character==s[10]{
		f=font[10]
	}
	return f
}
//in: string "07:35:43"
//out: ["0","7",":","3",....}
func string_strings_of_array(str string)[]string{
	arr:=strings.Split(str,"")
	return arr
}
// in : string_hhmmss
// out: fonts_array
func string_hhmmss_to_fonts_in_array(str_arr []string)[][]string{
	var font_string [][]string
	for i:=0;i<8;i++{
		str_ar:=str_arr[i]
		font_str:=convert_1_character_to_1_font(str_ar,font)
		font_string=append(font_string,font_str)
	}
	return font_string
}
//in:font1 []string ,font2 []string
//out: font []string
func merge_2_font(font1 []string,font2 []string)[]string{
	var fontnew=[]string {"","",""}
	for i:=0;i<(len(font1));i++{
		fontnew[i]=font1[i]+font2[i]
	}
	return fontnew
}
func merge_array_of_fonts(array_of_fonts [][]string)[]string{
	var merged_font=[]string{"","",""}
	for i:=0;i<(len(array_of_fonts));i++{
		merged_font=merge_2_font(merged_font,array_of_fonts[i])
	}
	return merged_font
} 
func clear_screen(){
	cmd := exec.Command("clear")
    cmd.Stdout = os.Stdout
    cmd.Run()
}
func display(array []string){
	for i:=0;i<(len(array));i++{
		row_i:=array[i]
		fmt.Println(row_i)
	}
}
//in :	
//
func main(){
	pre_sec:=0
	for true{
		t:=time.Now()
		s:=t.Second()
		if s!=pre_sec{
			clear_screen()
			str:=read_system_clock()
			ar:=string_strings_of_array(str)
			font_hhmmss:=string_hhmmss_to_fonts_in_array(ar)
			a:=merge_array_of_fonts(font_hhmmss)
			display(a)
		}
		pre_sec=s
		time.Sleep((1/1000)*time.Second)
	}
}
